python/lsst/ts/m2com/utility.py

```diff
@@ -444,3 +447,25 @@ def select_axial_hardpoints(
     hardpoints.sort()
 
     return hardpoints
+
+
+async def cancel_task_and_wait(task: asyncio.Task | asyncio.Future) -> None:
+    """Cancel the asynchronous task (if not done) and wait for it to complete.
+
+    Parameters
+    ----------
+    task : `asyncio.Task` or `asyncio.Future`
+        Asynchronous task.
+    """
+
+    if not task.done():
+        task.cancel()
+
+        # This might be a "Future" object. Check it is awaitable or not.
+        if is_coroutine(task):
```

Comment:
So, you are not trusting `mypy` type check to check this for you? Also, if this is false, you will most likely never get here as calling `task.done()` above will raise an exception if this is not a coroutine. Maybe nest these two `if` statements together, e.g.?


```
if is_coroutine(task) and not task.done():
    ....
```

Reply:
I need to do this because although the `Future` and `Task` objects can be cancelled, only the `Task` object is await-able with the `asyncio.CancelledError`.

If we do the following when the `task` is a `Future` object:

```
if is_coroutine(task) and not task.done():
    print("condition 1")
    task.cancel()
else:
    print("condition 2")
```

We will always get the condition 2 and can't do the `task.cancel()`.

Thanks!